#Since my 3 replicates had such low sequencing depth, I called peaks by merging 

library(rtracklayer)
library(GenomicRanges)

#Import peaks list generated by MACS2
consensus_peaks=import('/Volumes/SEK_Drive/ChIP-Seq/CicsfGFP_Replicates_Analysis/Merged_Peak_Calling/CicsfGFP_to_OreR_q0.01_nomodel/CicsfGFP_to_OreR_q0.01_nomodel_peaks.narrowPeak')
consensus_peaks=consensus_peaks[order(consensus_peaks$qValue,decreasing=TRUE)]

#Keep only the main chromosomes
seqlevels(consensus_peaks,pruning.mode='coarse')=c('chr2L','chr2R','chr3L','chr3R','chr4','chrX')

#Remove Dsor1 and Rl contamination and MtnA and regions of the X chromosome
g=GRanges(seqnames=c('chrX','chr2R','chr3R','chrX','chrX'),IRanges(start=c(9246668,1062236,9779069,23363605,2244894),end=c(9250711,1128431,9788708,23503104,2248904)))
g1=findOverlaps(g,consensus_peaks)
consensus_peaks=consensus_peaks[-unique(subjectHits(g1))]

#Convert peaks to be able to view in the Genome browser

raw = read.table('/Volumes/SEK_Drive/ChIP-Seq/CicsfGFP_Replicates_Analysis/Merged_Peak_Calling/CicsfGFP_to_OreR_q0.01_nomodel/CicsfGFP_to_OreR_q0.01_nomodel_peaks.narrowPeak', header = FALSE)
raw.g=GRanges(raw[,1], IRanges(raw[,2], raw[,3]))
colnames(raw) = c('chr','start','stop','name','score','strand','enrichment','p.val','q.val','summit_rel_to_start')
raw=raw[unique(subjectHits(findOverlaps(consensus_peaks,raw.g))),]
write.table(raw[,1:6], '/Volumes/SEK_Drive/ChIP-Seq/CicsfGFP_Replicates_Analysis/CicsfGFP_to_OreR_q0.01_nomodel.narrowPeak.UCSC', quote = FALSE, sep = '\t', row.names = FALSE, col.names = FALSE)

## Time to annotate our consensus peaks
library(ChIPseeker)
library(GenomicRanges)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene) 
txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
library (org.Dm.eg.db) 

#Annotate Peaks
peakAnno <- annotatePeak(consensus_peaks, tssRegion=c(-500, 500),genomicAnnotationPriority = c('Promoter','5UTR','Intron','Exon','3UTR',"Downstream","Intergenic"),TxDb=txdb, annoDb="org.Dm.eg.db",addFlankGeneInfo = TRUE)
peakAnno.df <- as.data.frame (peakAnno) # convert to dataframe

#Try to fill in the annotations that have NA
library(AnnotationDbi)

w=which(is.na(peakAnno.df$SYMBOL))
for (i in 1:length(w)) {
  if(!is.na(peakAnno.df[w[i],24])) {
    p=peakAnno.df[w[i],24:26]
    blah=data.frame(strsplit(p[,2],';'),strsplit(p[,1],';'),strsplit(p[,3],';'),stringsAsFactors=FALSE)
    blah[,4]=mapIds(org.Dm.eg.db,keys=as.character(blah[,1]),column='SYMBOL',keytype='ENSEMBL',multiVals='first')
    blah[,5]=mapIds(org.Dm.eg.db,keys=as.character(blah[,1]),column='GENENAME',keytype='ENSEMBL',multiVals='first')
    colnames(blah)=colnames(peakAnno.df[c(18,19,20,22,23)])
    peakAnno.df[w[i],c(12,13,14,15,16,17,21)]=NA
    peakAnno.df[w[i],c(18,19,20,22,23)]=blah[1,]
    blah=0
  } else {
    peakAnno.df[w[i],]=peakAnno.df[w[i],]
  }
}

#Let's only look at peaks with a log fold change >5

s=which(consensus_peaks$signalValue<5)
peakAnno.df=peakAnno.df[-s,]

#save your annotates peaks as a granges object
consensus_peaks=GRanges(peakAnno.df)
save(consensus_peaks, file = '/Volumes/SEK_Drive/ChIP-Seq/CicsfGFP_Replicates_Analysis/consensus_peaks_cictoOreR_q0.01.gr')

#Here are some other graphs we can look at
covplot(consensus_peaks)

promoter <- getPromoters(TxDb=txdb, upstream=500, downstream=500)
tagMatrix <- getTagMatrix(consensus_peaks, windows=promoter, weightCol="score")
tagHeatmap(tagMatrix, xlim=c(-500, 500), color="darkblue")

plotAvgProf(tagMatrix, xlim=c(-500, 500), xlab="Position", ylab = "Read Count Frequency", conf=0.95, resample=1000) # TSS profile with 95% CI

#visualize the annotations
plotAnnoPie(annotatePeak(consensus_peaks, tssRegion=c(-500, 500),genomicAnnotationPriority = c('Promoter','5UTR','Intron','Exon','3UTR',"Downstream","Intergenic"),TxDb=txdb, annoDb="org.Dm.eg.db",addFlankGeneInfo = TRUE)
)

library(ReactomePA)
pathway.enrich <- enrichPathway(peakAnno.df$ENTREZID, organism="fly", readable=T)
dotplot(pathway.enrich)

library(clusterProfiler)
pe=enrichGO(peakAnno.df$ENTREZID,OrgDb="org.Dm.eg.db", readable=T)#18 enriched terms- DNA binding, RNA pol II specific, etc...
dotplot(pe)
